crates/bevy_animation/src/animation_curves.rs

```diff
@@ -429,63 +784,319 @@ where
+            stack_blend_weights_and_graph_nodes: vec![],
+            blend_register_morph_target_weights: vec![],
+            blend_register_blend_weight: None,
+            morph_target_count: None,
+            phantom: PhantomData::<C>,
+        })
+    }
+}
+
+impl<C> AnimationCurveEvaluator for WeightsCurveEvaluator<C>
+where
+    C: IterableCurve<f32> + Debug + Clone + Reflectable,
+{
+    fn blend(&mut self, graph_node: AnimationNodeIndex) -> Result<(), AnimationEvaluationError> {
+        let Some(&(_, top_graph_node)) = self.stack_blend_weights_and_graph_nodes.last() else {
+            return Ok(());
+        };
+        if top_graph_node != graph_node {
+            return Ok(());
+        }
+
+        let (weight_to_blend, _) = self.stack_blend_weights_and_graph_nodes.pop().unwrap();
+        let stack_iter = self.stack_morph_target_weights.drain(
+            (self.stack_morph_target_weights.len() - self.morph_target_count.unwrap() as usize)..,
+        );
+
+        match self.blend_register_blend_weight {
+            None => {
+                self.blend_register_blend_weight = Some(weight_to_blend);
+                self.blend_register_morph_target_weights.clear();
+                self.blend_register_morph_target_weights.extend(stack_iter);
+            }
+
+            Some(ref mut current_weight) => {
+                *current_weight += weight_to_blend;
+                for (dest, src) in self
+                    .blend_register_morph_target_weights
+                    .iter_mut()
+                    .zip(stack_iter)
+                {
+                    *dest = f32::interpolate(dest, &src, weight_to_blend / *current_weight);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn push_blend_register(
+        &mut self,
+        weight: f32,
+        graph_node: AnimationNodeIndex,
+    ) -> Result<(), AnimationEvaluationError> {
+        if self.blend_register_blend_weight.take().is_some() {
+            self.stack_morph_target_weights
+                .append(&mut self.blend_register_morph_target_weights);
+            self.stack_blend_weights_and_graph_nodes
+                .push((weight, graph_node));
+        }
+        Ok(())
+    }
+
+    fn commit<'a>(
+        &mut self,
+        _: Option<Mut<'a, Transform>>,
+        mut entity: AnimationEntityMut<'a>,
+    ) -> Result<(), AnimationEvaluationError> {
+        if self.stack_morph_target_weights.is_empty() {
+            return Ok(());
+        }
+
+        for (dest, src) in entity
+            .get_mut::<MorphWeights>()
+            .ok_or_else(|| {
+                AnimationEvaluationError::ComponentNotPresent(TypeId::of::<MorphWeights>())
+            })?
+            .weights_mut()
+            .iter_mut()
+            .zip(
+                self.stack_morph_target_weights[(self.stack_morph_target_weights.len()
+                    - self.morph_target_count.unwrap() as usize)..]
+                    .iter(),
+            )
```

Comment:
It's somewhat hard to follow what's going on here. Could the index be split into its own variable?

Reply:
Done.